Fill in validity circuit (#67)
junyechen1996/draft-chen-cfrg-vdaf-pine: draft-chen-cfrg-vdaf-pine.md

  each wraparound check, and the success bits in wraparound checks.
* The last `num_wr_checks` are the wraparound check results, i.e., the dot
  products of the encoded gradient and the random vectors.

It also unpacks the "joint randomness" that is shared between the Client and
Aggregators, to compute random linear combinations of all the checks:

* The first joint randomness field element is to reduce over the bit checks at
  all bit entries.
* The second joint randomness field element is to reduce over all the quadratic
  checks in wraparound check.
* The last joint randomness field element is to reduce over all the checks,
  which include the reduced bit check result, the L2 norm equality check, the
  L2 norm range check, the reduced quadratic checks in wraparound check, and
  the success count check for wraparound check.

In the subsections below, we outline the various checks computed by the validity
circuit, which includes the bit check on all the bit entries
{{valid-bit-check}}, the L2 norm check {{valid-norm-check}}, and the wraparound
check {{valid-wr-check}}. Some of the auxiliary functions in these checks are
defined in {{pine-auxiliary}}.

> ### Range Check {#range-check}

Reviewer: Perfect!

A common subroutine used in the validity circuit is the "range check", i.e.,
checking if a `value` is in the range of `[B1, B2]`, over the field size `q`
(See Figure 1 in {{ROCT23}}). The Client computes the "`v` bits", the bit
representation of `value - B1` (modulo `q`), and the "`u` bits", the bit
representation of `B2 - value` (modulo `q`). Assuming the Aggregators have

> verified the `v` bits and `u` bits are indeed composed of bits, the Aggregators

Reviewer: ```suggestion
verified the `v` bits and `u` bits are indeed composed of bits (as described in {{valid-bit-check}}), the Aggregators
```

can verify `value` is in the desired range, by checking the decoded value from
the `v` bits, and the decode value from the `u` bits sum up to `B2 - B1`
(modulo `q`).

As an optimization for communication cost per Remark 3.2 in {{ROCT23}}, the
Client can skip sending the `u` bits if `B2 - B1 + 1` (modulo `q`) is a power of

> 2. This is because the available `v` bits can naturally bound `value - B1` to be

Reviewer: ```suggestion
`2`. This is because the available `v` bits can naturally bound `value - B1` to be
```

## The FLP Circuit

The validity circuit unpacks the encoded measurement as the following:

* The first `dimension` entries are the `encoded_gradient`, the field elements
  encoded from the floating point numbers.
* The next `bit_checked_len` entries are expected to be bits, and should contain
  the bits from L2-norm check, the bits from the range check in wraparound

>   checks, and the success bits in wraparound checks.

Author: I quickly mentioned it, but also referred to {{encode-wr-checks}}, which has yet to be filled.

>   checks, and the success bits in wraparound checks.

Reviewer: Once you're satisfied it's defined, feel free to resolve this.

* The last `num_wr_checks` are the wraparound check results, i.e., the dot
  products of the encoded gradient and the random vectors.

It also unpacks the joint randomness to compute random linear combinations of
all the checks:

* The first joint randomness field element is to reduce over the bit checks at
  all bit entries.
* The second joint randomness field element is to reduce over all the quadratic
  checks in wraparound check.
* The last joint randomness field element is to reduce over all the checks,
  which include the reduced bit check result, the L2-norm equality check, the

>   L2-norm range check, the reduced quadratic checks in wraparound check, and

Author: I couldn't really come up with a better name than "quadratic check"... So I used the term in the paper.

>   L2-norm range check, the reduced quadratic checks in wraparound check, and

Reviewer: That's fine for now, I think connecting it to the paper is the main thing we need.

  the success count check for wraparound check.

Some of the auxiliary functions are defined in {{pine-auxiliary}}.

> 

Reviewer: The jump  to the next subusbusections is rather abrupt. We need some fence posting here to let the reader know how the remainder is organized.

### Bit Check {#valid-bit-check}

> 

Reviewer: Before we get to the technical details, we need to first say what the purpose is of checking that `bit_checked` is made up of bits.

The bit check computes a random linear combination of the polynomial `x(x-1)`

> evaluated at all bit entries of `bit_checked`, with the coefficients computed

Reviewer: ```suggestion
evaluated at each element of `bit_checked`, with the coefficients computed
```

> evaluated at all bit entries of `bit_checked`, with the coefficients computed

Reviewer: I don't think you mean "coefficients" here. We're evaluating a polynomial $p(r) = c_0 + c_1 r + c_1 + r^2 + \cdots$ at a random point $r$: what are the ceofficients?

from the joint randomness `r_bit_check`. Note we need to normalize the constant
`1` in the polynomial, by multiplying it by the inverse of number of shares

> `shares_inv`.

Reviewer: I'd say let's save this note until we've handled the TODO below.

In fact, it may be worth adding an in line comment somewhere in `eval()` explaining the purpose of `shares_inv`.

> TODO Put `PineValid.eval_bit_check()` implementation here.

### L2-Norm Check {#valid-norm-check}

> The validity circuit verifies two properties for L2-norm check:

Reviewer: ```suggestion
The validity circuit verifies two properties of the L2 norm reported by the Client:
```

* Equality check: The computed sum of squares of `encoded_gradient` is equal to

>   the squared L2-norm decoded from the bit representation `sq_norm_v_bits`

Reviewer: I don't think variable `sq_norm_v_bits` has not been defined? Ask yourself if you actually need to define it in order to convey what's going on.

I think we should try to explain what's going on without referring to variables in the circuit. What would help is an explanation of how range checking works in PINE. That way all you need to do here is say that the circuit decodes the bit-encoding of the reported norm and performs the range check.

Likewise for "wraparound check" section below. the explanation of how range checking works should go somewhere towards the top of the "FLP Circuit" section.

Reviewer: The explanation is great. I think we can now simplify this text accordingly, by not mentioning variables. Also, I suggest describing what the circuit actuall does here:

> Equality check: The squared norm of the gradient is equal to the value reported by the Client. For this, the Aggregators decode their shares of the bit-representation of the squared norm (as defined above in {{valid-bit-check}}), compute their shares of the squared norm from their shares of the gradient, and check that the values are equal.
> Range check: The squared norm reported by the Client is in the desired range. For this, the Aggregators run the range check described in {{range-check}}.

Likewise for the other "sub-checks".
